### PR TITLE
Numeric opcodes

### PR DESCRIPTION
**Pull request by ehildenb**
adding numeric opcode semantics, not ready for review

### CODE REVIEW

File: tests/interactive/arithmetic.wast [hunk 1]
```diff
@@ -9,3 +9,85 @@ i32.sub
...
+i64.const #pow63
+i64.eq
+
+i64.const #unsigned(i64, #signed(i64, #pow64 -Int 1))
+i64.const #pow64 -Int 1
+i64.eq
+
+i32.const 5
+i32.const #pow31 +Int 15
+i32.div_s
```
Comment (pirapira) [inline tests/interactive/arithmetic.wast]: My Haskell interpreter says
```
Prelude> (15 + (2 ^ 31) - (2 ^ 32) ) `div` 5 + 2 ^ 32
3865470569
```
and `div` in Haskell truncates towards zero (as `div_s` in webassembly)
```
Prelude> -4 `div` 5
0
```

**Comment by ehildenb** [comment]
@pirapira fixed, `git fetch ; git diff --word-diff origin/numeric-opcodes numeric-opcodes` shows the difference. Had to trade `/Int` for `divInt` (which rounds towards zero).

File: wasm-syntax.md [hunk 2]
```diff
@@ -141,9 +187,83 @@ A large portion of the available opcodes are pure arithmetic.
...
+    rule #evalIBinOp(TYPE . rem_u, I1, I2) => < TYPE > (I1 %Int I2) requires I2 =/=Int 0
+
+    syntax IBinOp ::= "div_s" | "rem_s"
  // -----------------------------------
+    rule #evalIBinOp(i32 . div_s, I1, I2) => < i32 > #unsigned(i32, #signed(i32, I1) divInt #signed(i32, I2))
+      requires I2 =/=Int 0 andBool (I1 =/=Int #pow31 orBool (I2 =/=Int #pow32 -Int 1))
+    rule #evalIBinOp(i64 . div_s, I1, I2) => < i64 > #unsigned(i64, #signed(i64, I1) divInt #signed(i64, I2))
+      requires I2 =/=Int 0 andBool (I1 =/=Int #pow63 orBool (I2 =/=Int #pow64 -Int 1))
+
+    rule #evalIBinOp(TYPE . rem_s, I1, I2) => < TYPE > #unsigned(TYPE, #signed(TYPE, I1) modInt #signed(TYPE, I2))
```
Comment (pirapira) [inline wasm-syntax.md]: `div_s` is t-division, so `rem_s` should be t-division too, I guess.

File: tests/interactive/arithmetic.wast [hunk 3]
```diff
@@ -9,3 +9,85 @@ i32.sub
...
+
+i32.const 7
+i32.const #pow31 +Int 15
+i32.rem_s
```
Comment (pirapira) [inline tests/interactive/arithmetic.wast]: The result's sign should be the sign of the dividend.  https://webassembly.github.io/spec/core/exec/numerics.html#op-irem-s

File: tests/stack/arithmetic.wast [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+i32.const 2
+i32.const 15
+i32.rem_u
+#assertTopStack < i32 > 1 "rem_u2"
+
+i32.const 5
+i32.const #pow31 +Int 15
+i32.div_s
+#assertTopStack < i32 > 0 "div_s"
```
Comment (ehildenb) [inline tests/stack/arithmetic.wast]: @pirapira see interaction with `bc` (command line calculator):

```
scale = 4

n = (15 + (2 ^ 31))
d = 5

signed = n - (2 ^ 32)
dived = signed / d
dived

trunced = -429496726
trunced + (2 ^ 32)
```

Produces output:

```
-429496726.6000
3865470570
```

Not sure what's going on with GHCi, but I think that it should be 386547050 (currently asserting 0 in test until this is cleared up).

Comment (pirapira) [reply tests/stack/arithmetic.wast]: You're right.  GHCi behaves weird.

Sane on small numbers
```
Prelude> -4 `div` 5
0
Prelude> -5 `div` 5
-1
```

Differently on bigger numbers
```
Prelude> (15 + (2 ^ 31) - (2 ^ 32) ) / 5.0
-4.294967266e8
Prelude> (15 + (2 ^ 31) - (2 ^ 32) ) `div` 5
-429496727
```

Comment (pirapira) [reply tests/stack/arithmetic.wast]: Now i understand what happned: https://ghc.haskell.org/trac/ghc/ticket/643

```
-4 `div` 5
```
is 
```
-(4 `div` 5)
```

File: wasm-syntax.md (hunk 2 above)
Comment (pirapira) [reply wasm-syntax.md]: Now it looks like both are e-division.

**Pull request merged**